[InvertedIndex](Variant) support inverted index for array type in variant

eldenmoon opened the pull request:
### What problem does this PR solve?
This pull request includes several changes to support inverted indexes for arrays and improve the handling of complex data types in the codebase. The most important changes include modifications to the inverted index writer, updates to expression evaluation, and the addition of new regression tests.

### Improvements to inverted index support:

* [`be/src/olap/rowset/segment_v2/inverted_index_writer.cpp`](diffhunk://#diff-97781916b276f771710ab520c79ca29d5e4e331296fad7573fc9933a376dc165L79-R91): Updated `check_support_inverted_index` method to support one-level arrays and removed the check for extracted columns.
* [`be/src/vec/exprs/vexpr.cpp`](diffhunk://#diff-46b3e9eb194dc83ddbdb52e26fc18f9204669761f0dc6f90502dc2b83828cb77L656-R656): Modified `_evaluate_inverted_index` method to remove nullable types from the target type during evaluation.

### Addition of regression tests:

* [`regression-test/data/variant_p0/with_index/test_array_inverted_index_profile.out`](diffhunk://#diff-ee440c0dc7797173bf524be705b22d7ab4044808e19d81825e8ab79b080ad9fcR1-R51): Added new test cases for array inverted index profiles.
* [`regression-test/suites/variant_p0/with_index/test_array_contains_with_inverted_index.groovy`](diffhunk://#diff-4fcc39383edd29ad28d9830c401188206c36b23949f831c6cbceb53309ca8b21R1-R112): Introduced a new test suite to validate the `array_contains` function with inverted indexes.
* [`regression-test/suites/variant_p0/with_index/test_array_index_write.groovy`](diffhunk://#diff-1658d944fdb787eadefa6eaff841326f6f68c0fccf58899e9a6a0c631017ab63R1-R93): Added tests to verify writing to tables with array inverted indexes and handling various data types.
Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

eldenmoon committed:
[InvertedIndex](Variant) supoort inverted index for array type in variant

hello-stephen commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

eldenmoon commented:
run buildall

eldenmoon commented:
This pull request includes several changes to support inverted indexes for arrays and improve the handling of complex data types in the codebase. The most important changes include modifications to the inverted index writer, updates to expression evaluation, and the addition of new regression tests.

### Improvements to inverted index support:

* [`be/src/olap/rowset/segment_v2/inverted_index_writer.cpp`](diffhunk://#diff-97781916b276f771710ab520c79ca29d5e4e331296fad7573fc9933a376dc165L79-R91): Updated `check_support_inverted_index` method to support one-level arrays and removed the check for extracted columns.
* [`be/src/vec/exprs/vexpr.cpp`](diffhunk://#diff-46b3e9eb194dc83ddbdb52e26fc18f9204669761f0dc6f90502dc2b83828cb77L656-R656): Modified `_evaluate_inverted_index` method to remove nullable types from the target type during evaluation.

### Addition of regression tests:

* [`regression-test/data/variant_p0/with_index/test_array_inverted_index_profile.out`](diffhunk://#diff-ee440c0dc7797173bf524be705b22d7ab4044808e19d81825e8ab79b080ad9fcR1-R51): Added new test cases for array inverted index profiles.
* [`regression-test/suites/variant_p0/with_index/test_array_contains_with_inverted_index.groovy`](diffhunk://#diff-4fcc39383edd29ad28d9830c401188206c36b23949f831c6cbceb53309ca8b21R1-R112): Introduced a new test suite to validate the `array_contains` function with inverted indexes.
* [`regression-test/suites/variant_p0/with_index/test_array_index_write.groovy`](diffhunk://#diff-1658d944fdb787eadefa6eaff841326f6f68c0fccf58899e9a6a0c631017ab63R1-R93): Added tests to verify writing to tables with array inverted indexes and handling various data types.

eldenmoon commented:
run buildall

xiaokang commented on be/src/vec/exprs/vexpr.cpp:
```diff
@@ -653,7 +653,7 @@ Status VExpr::_evaluate_inverted_index(VExprContext* context, const FunctionBase
                                 ->get_storage_name_and_type_by_column_id(column_id);
                 auto storage_type = remove_nullable(storage_name_type->second);
-                auto target_type = cast_expr->get_target_type();
+                auto target_type = remove_nullable(cast_expr->get_target_type());
```
Why add remove_nullable?

eldenmoon replied:
remove null to get real type, it's a bug in original code

xiaokang commented on be/test/common/schema_util_test.cpp:
```diff
@@ -48,6 +48,18 @@ void construct_subcolumn(TabletSchemaSPtr schema, const FieldType& type, int32_t
+    if (type == FieldType::OLAP_FIELD_TYPE_ARRAY) {
+        TabletColumn array_item_col;
+        // double not support inverted index
+        array_item_col.set_type(FieldType::OLAP_FIELD_TYPE_DOUBLE);
```
Is it a trick to add a default type DOUBLE that does not support inverted index?

xiaokang replied:
Where do you set the real item type and why not set the real item type here?

eldenmoon replied:
it's ut

xiaokang commented on regression-test/suites/variant_github_events_new_p2/load.groovy:
```diff
@@ -128,4 +128,26 @@ suite("regression_test_variant_github_events_p2", "nonConcurrent,p2"){
+    profile("test_profile_1") {
+        sql """ set enable_common_expr_pushdown = true; """
+        sql """ set enable_common_expr_pushdown_for_inverted_index = true; """
+        sql """ set enable_pipeline_x_engine = true;"""
```
Is enable_pipeline_x_engine necessary? Setting it explicitly will prevent session var fuzzy.

eldenmoon replied:
profile needed

xiaokang commented on regression-test/suites/variant_p0/with_index/test_array_contains_with_inverted_index.groovy:
```diff
@@ -0,0 +1,112 @@
+    // If we use common expr pass to inverted index , we should set enable_common_expr_pushdown = true
+    sql """ set enable_common_expr_pushdown = true; """
+    sql """ set enable_common_expr_pushdown_for_inverted_index = true; """
+    sql """ set enable_pipeline_x_engine = true;"""
```
dito

eldenmoon replied:
done

xiaokang commented on regression-test/suites/variant_p0/with_index/test_array_contains_with_inverted_index.groovy:
```diff
@@ -0,0 +1,112 @@
+        order_qt_sql """ select * from tai where (array_contains(cast(inventors['inventors'] as array<text>), ${p}) and apply_date = '2017-01-01') or apply_date = '2019-01-01' order by id; """
+    }
+
+    // test arrays_overlap with inverted index
```
How to check whether inverted index is used?

eldenmoon replied:
debug point checked

eldenmoon committed:
fix test

eldenmoon commented:
run buildall

eldenmoon commented:
run buildall

airborne12 reviewed (approved):
LGTM

xiaokang reviewed (approved):
LGTM
